src/Data.Json.JTable.Internal.purs

```diff
@@ -0,0 +1,214 @@
+module Data.Json.JTable.Internal
+  ( renderJTableRaw, renderRows, renderThead, renderTbody, tsToRows, sortTree
+  , Tree(..), tPath, tWidth, tHeight, tKids
+  , Cell(..), cCursor, cWidth, cHeight, cJsonPrim
+  , JPath(..), Table(..)
+  , tFromJson, tMergeArray, widthOfPrimTuple, insertHeaderCells
+  , cFromJson, cMergeObj, mergeObjTuple
+  , _cN, toPrim, strcmp, _nattr, _cspan, _rspan
+  ) where
+
+import Math (max)
+import Data.Either
+import Data.Maybe (Maybe(..), fromMaybe, isJust)
+import Data.Tuple
+import Data.String (joinWith, localeCompare)
+import Data.Array
+import Data.Foldable
+import qualified Data.Array.Unsafe as AU
+import qualified Data.StrMap as M
+import Data.Foldable (foldl, any, all, mconcat, elem)
+import Data.Traversable (for)
+import Control.Alt
+import Control.Apply
+import Control.Bind
+import Control.Monad
+import Control.MonadPlus
+
+import Data.Argonaut.Core
+import Data.Argonaut.JCursor
+import Data.Argonaut.Parser (jsonParser)
+import Text.Smolder.HTML (thead, tbody, tr, th, td)
+import Text.Smolder.Markup (Markup(..), Attribute(..), attribute, (!))
+
+-- path of object keys, with array indices omitted
+type JPath = [String]
+
+-- rows of cells
+type Table = [[Cell]]
+
+-- header data
+data Tree = T JPath Number Number [Tree]
+tPath (T p _ _ _) = p
+tWidth (T _ w _ _) = w
+tHeight (T _ _ h _) = h
+tKids (T _ _ _ k) = k
+
+instance showTree :: Show Tree where
+  show (T p w h k) = joinWith " " ["(T", show p, show w, show h, show k, ")"]
+
+-- cell data
+data Cell = C  JCursor Number Number JsonPrim
+cCursor (C c _ _ _) = c
+cWidth (C _ w _ _) = w
+cHeight (C _ _ h _) = h
+cJsonPrim (C _ _ _ j) = j
+
+instance showCell :: Show Cell where
+  show (C c w h j) = joinWith " " ["(C", show w, show h, show c, show j, ")"]
+
+
+foreign import jnull "var jnull = null;" :: Json
+
+_cN = const primNull
+toPrim = (foldJson _cN primBool primNum primStr _cN _cN) :: Json -> JsonPrim
+zipWithIndex xs = zip xs (0 .. length xs - 1) 
+
+so :: forall a. Boolean -> a -> Maybe a
+so b x = if b then Just x else Nothing
+
+
+-- maybe return the width of a tuple composed of primitive values
+widthOfPrimTuple :: JPath -> [Json] -> Maybe Number
+widthOfPrimTuple path ja = 
+  path !! 0 *> ja !! 1 *> let
```

Comment:
1. We don't know if that is the case. The function returns Nothing if there are non-primitives.
2. Can you give an example? Actually if all elements have the same type, it's an array (rendered vertically), and if elements have different types, then it's a tuple (rendered horizontally)
3. how? `foldJson (const 0) (const 1) (const 2) (const 3) (const 4) (const 4)` doesn't look much simpler.